CONTRIBUTING.md

```diff
@@ -1,16 +1,721 @@
+    size = 10
+  }
+}
+```
+
+Where this results in objects with too many attributes, we usually split attributes between required and optional by adding a second level, as in this example where VM `attached_disks[].options` contains less used attributes and can be set to null if not needed.
+
+```hcl
+module "simple-vm-example" {
+  source     = "./modules/compute-vm"
+  project_id = var.project_id
+  zone       = "europe-west1-b"
+  name       = "test"
+  attached_disks = [
+    { name="data", size=10, source=null, source_type=null, options=null }
+  ]
+}
+```
+
+Whenever options are not passed like in the example above, we typically infer their values from a defaults variable which can be customized when using defaults across several items. In the following example instead of specifying regional PD options for both disks, we set their options to `null` and change the defaults used for all disks.
+
+```hcl
+module "simple-vm-example" {
+  source     = "./modules/compute-vm"
+  project_id = var.project_id
+  zone       = "europe-west1-b"
+  name       = "test"
+  attached_disk_defaults = {
+    auto_delete = false
+    mode = "READ_WRITE"
+    replica_zone = "europe-west1-c"
+    type = "pd-balanced"
+  }
+  attached_disks = [
+    { name="data1", size=10, source=null, source_type=null, options=null },
+    { name="data2", size=10, source=null, source_type=null, options=null }
+  ]
+}
+```
+
+#### Depend outputs on internal resources
+
+We mentioned this principle when discussing encapsulation above but it's worth repating it explicitly: **set explicit dependencies in outputs so consumers will wait for full resource configuration**.
+
+As an example, users can safely reference the project module's `project_id` output from other modules, knowing that the dependency tree for project configurations (service activation, IAM, etc.) has already been defined inside the module itself. In this particular example the output is also interpolated instead of derived from the resource, so as to avoid issues when used in `for_each` keys.
+
+```hcl
+output "project_id" {
+  description = "Project id."
+  value       = "${local.prefix}${var.name}"
+  depends_on = [
+    google_project.project,
+    data.google_project.project,
+    google_project_organization_policy.boolean,
+    google_project_organization_policy.list,
+    google_project_service.project_services,
+    google_compute_shared_vpc_service_project.service_projects,
+    google_project_iam_member.shared_vpc_host_robots,
+    google_kms_crypto_key_iam_member.service_identity_cmek,
+    google_project_service_identity.servicenetworking,
+    google_project_iam_member.servicenetworking
+  ]
+}
+```
+
+#### Why we don't use random strings in names
+
+This is more a convention than a design principle, but it's still important enough to be mentioned here: we **never use random resources for naming** and instead rely on an optional `prefix` variable which is implemented in most modules.
```

Comment:
s/we never use random resources for naming/we never use strings for resources naming/